### PR TITLE
Added missing type names to uri parsing regex for memory provider

### LINKED ISSUE #44087: Duplicate Layer changes Field type and name
Imported a CSV and then copied the data to a new layer to be able to edit it. After that, I duplicated the copied layer, as I needed different visual representations. The duplicated layer has an issue where the first field "Pozo", type Qstring text, is duplicated as "Pozo:text(0,0)" type Qstring string, which creates some issues with symbology filters already in place from the original layer.

This is the source layer (copied layer from CSV layer):
Original

This is the duplicate layer (duplicated from the copied layer):
Duplicate

This is the dataset:
[BP-004.csv](https://github.com/qgis/QGIS/files/6785245/BP-004.csv)

This are the import settings I used:
Import

My system is a Macbook Pro Retina 2013 with macOS Big Sur 11.4 and QGIS info is the following:

Info

Steps I followed:

1. Imported dataset
2. Selected layer, selected all, and copied with CMD+A followed by CMD+C
3. Pasted copied data onto a new temporary layer with OPTION+CMD+V (this temporary layer has the correct field name)
4. Duplicated the newly created layer by right clicking and selecting "duplicate layer" (this duplicate layer has the incorrect field name)

I hope that's enough info. Thank you in advance.

### PR DESCRIPTION
**Pull request by domi4484**
- This was causing strange field names when cloning layers. fixes #44087
- Some code cleanup + QRegexp replaced with QRegularExpression

### CODE REVIEW

File: src/core/providers/memory/qgsmemoryprovider.cpp [hunk 1]
```diff
@@ -118,96 +118,108 @@ QgsMemoryProvider::QgsMemoryProvider( const QString &uri, const ProviderOptions
...
       QVariant::Type type = QVariant::String;
       QVariant::Type subType = QVariant::Invalid;
-      QString typeName( QStringLiteral( "string" ) );
-      int length = 255;
-      int precision = 0;
+      QString typeName( regularExpressionMatch.captured( 1 ).toLower() );
 
-      int pos = reFieldDef.indexIn( name );
-      if ( pos >= 0 )
+      // Search typname correspondance in native types
```
Comment (agiudiceandrea) [inline src/core/providers/memory/qgsmemoryprovider.cpp:156]:
```suggestion
      // Search typeName correspondence in native types
```

**Comment by agiudiceandrea** [comment]
The tests are failing because it seems that the field type is optional and if it is not specified in the uri string, then the field type is expected to default to a string field type:
https://github.com/qgis/QGIS-Documentation/blame/master/docs/pyqgis_developer_cookbook/vector.rst#L953-L955
> field=name:type(length,precision)
> Specifies an attribute of the layer. The attribute has a name, and optionally a type (integer, double, or string), length, and precision.

https://github.com/qgis/QGIS/blob/92fc43505de8b1a781834380bbc1a8cf61c84292/src/core/providers/memory/qgsmemoryprovider.cpp#L134

**Comment by domi4484** [comment]
@agiudiceandrea Thank you for the analysis of the failing test! I changed it back to the expected behavior when no type is specified.
Additionally I have made the regexp accepting unknown type names. If the type name is unknown the field is created as "string" and a warning is traced.

File: src/core/qgsfield.cpp [hunk 2]
```diff
@@ -312,7 +312,7 @@ QString QgsField::displayString( const QVariant &v ) const
       }
     }
     // Default for doubles with precision
-    else if ( d->type == QVariant::Double && d->precision > 0 )
+    else if ( d->precision > 0 )
```
Comment (nirvn) [inline src/core/qgsfield.cpp:315]: Why the change here?

**Review by nirvn — APPROVED** [review]
I like the added reliance on the native type when decoding fields from the URI string.

File: src/core/providers/memory/qgsmemoryprovider.cpp [hunk 3]
```diff
@@ -119,95 +119,101 @@ QgsMemoryProvider::QgsMemoryProvider( const QString &uri, const ProviderOptions
   if ( query.hasQueryItem( QStringLiteral( "field" ) ) )
   {
     QList<QgsField> attributes;
-    QRegExp reFieldDef( "\\:"
-                        "(int|integer|long|int8|real|double|string|date|time|datetime|binary|bool|boolean)" // type
-                        "(?:\\((\\-?\\d+)"                // length
-                        "(?:\\,(\\-?\\d+))?"                  // precision
-                        "\\))?(\\[\\])?"                  // array
-                        "$", Qt::CaseInsensitive );
+    QRegularExpression reFieldDef( "\\:"
```
Comment (nirvn) [inline src/core/providers/memory/qgsmemoryprovider.cpp:122]: @domi4484 , long hanging micro optimization possible here: `QRegularExpression -> const thread_local QRegularExpression`.

**Comment by agiudiceandrea** [comment]
@nirvn the issue fixed by this PR affects 3.16, too.

File: src/core/qgsfield.cpp (hunk 2 above)
Comment (domi4484) [reply src/core/qgsfield.cpp:315]: It is still in the same if of line 262

Comment (nirvn) [reply src/core/qgsfield.cpp:315]: Ok

**Comment by agiudiceandrea** [comment]
@domi4484 the automatic backports failed.

**Comment by domi4484** [comment]
Thanks for the reminder @agiudiceandrea, had almost forgotten
